pkg/controller/atlasinstance/atlasinstance_controller.go

```diff
@@ -0,0 +1,387 @@
+	ptr "k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	dbaasv1alpha1 "github.com/RHEcosystemAppEng/dbaas-operator/api/v1alpha1"
+	"go.mongodb.org/atlas/mongodbatlas"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/dbaas"
+	v1 "github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/project"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/provider"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/status"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/atlas"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/atlasinventory"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/watch"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/workflow"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/util/kube"
+)
+
+// MongoDBAtlasInstanceReconciler reconciles a MongoDBAtlasInstance object
+type MongoDBAtlasInstanceReconciler struct {
+	Client      client.Client
+	Clientset   kubernetes.Interface
+	AtlasClient *mongodbatlas.Client
+	watch.ResourceWatcher
+	Log             *zap.SugaredLogger
+	Scheme          *runtime.Scheme
+	AtlasDomain     string
+	GlobalAPISecret client.ObjectKey
+	EventRecorder   record.EventRecorder
+}
+
+type InstanceData struct {
+	ProjectName      string
+	ClusterName      string
+	ProviderName     string
+	RegionName       string
+	InstanceSizeName string
+}
+
+// Dev note: duplicate the permissions in both sections below to generate both Role and ClusterRoles
+
+// +kubebuilder:rbac:groups=dbaas.redhat.com,resources=mongodbatlasinstances,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=dbaas.redhat.com,resources=mongodbatlasinstances/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
+
+// +kubebuilder:rbac:groups=dbaas.redhat.com,namespace=default,resources=mongodbatlasinstances,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=dbaas.redhat.com,namespace=default,resources=mongodbatlasinstances/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",namespace=default,resources=secrets,verbs=get;list;watch
+
+func (r *MongoDBAtlasInstanceReconciler) Reconcile(cx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = cx
+	log := r.Log.With("MongoDBAtlasInstance", req.NamespacedName)
+	log.Info("Reconciling MongoDBAtlasInstance")
+
+	inst := &dbaas.MongoDBAtlasInstance{}
+	if err := r.Client.Get(cx, req.NamespacedName, inst); err != nil {
+		if apiErrors.IsNotFound(err) {
+			// CR deleted since request queued, child objects getting GC'd, no requeue
```

Comment:
How you are detecting the cluster when MongoDBAtlasInstance is deleted?

Reply:
When MongoDBAtlasInstance CR is deleted, the corresponding atlascluster CR (via owner references) is deleted. The cluster deletion is then done by atlascluster reconciler's Delete() method.